relay-metrics/src/aggregation.rs

```diff
@@ -1302,17 +1297,38 @@ impl Aggregator {
             return Err(AggregateMetricsErrorKind::InvalidStringLength.into());
         }
 
-        if !protocol::is_valid_mri(&key.metric_name) {
-            relay_log::debug!("invalid metric name {:?}", key.metric_name);
-            relay_log::configure_scope(|scope| {
-                scope.set_extra(
-                    "bucket.project_key",
-                    key.project_key.as_str().to_owned().into(),
-                );
-                scope.set_extra("bucket.metric_name", key.metric_name.into());
-            });
-            return Err(AggregateMetricsErrorKind::InvalidCharacters.into());
-        }
+        key.metric_name = match key.metric_name.parse::<MetricResourceIdentifier>() {
+            Ok(mri) => {
+                if matches!(mri.namespace, MetricNamespace::Unsupported) {
+                    relay_log::debug!("invalid metric namespace {:?}", key.metric_name);
+                    relay_log::configure_scope(|scope| {
+                        scope.set_extra(
+                            "bucket.project_key",
+                            key.project_key.as_str().to_owned().into(),
+                        );
+                        scope.set_extra("bucket.metric_name", key.metric_name.into());
+                    });
+                    return Err(AggregateMetricsErrorKind::InvalidNamespace.into());
+                }
+
+                let mut metric_name = mri.to_string();
+                // do this so cost tracking still works accurately.
+                metric_name.shrink_to_fit();
```

Comment:
Cost tracking uses `String::capacity()`, so it should still be accurate, but shrinking it sounds like a good idea anyway.

Reply:
I would not bother with such details here and rather ensure cost tracking is accurate enough (which it is). Otherwise we have to worry about that everywhere in the code base, which is not scalable.

Reply:
I dabbled around with it just now and I think it's more complicated to write a predictable testcase/assertion if shrink_to_fit is removed. Unless there are strong concerns about the perf overhead of doing this i would prefer to keep it.